README
NAME

    sixservo.ino - control 6 servo motors via serial console

SYNOPSIS

    after flashing into Arduino, connect via serial console 9600 to send
    commands:

        supported commands:
            sd [X] [int]             - turn servo to 0-180
            ss [X] [int]             - set puls width 500-2500
            sr [X] [int]             - set position relative to zero -1000-1000
            sweep [X] [int] [int]    - sweep servo [X] 0-180-0, optional count and step delay
            d [X]                    - detach / stop controlling the servo
            z [X]                    - set current as zero position
            c [X]                    - calibrate
            m [X] [int] [int]        - monitor servos, optional count and delay [s]
            save                     - save config
            dump [X]                 - show servo debug values
            ?                        - print this help
        [X] is servo index 1-6 or bitwise 000-077 (6 last bits as mask), default is zero -> 077
        ss 1 1500
        ok 1
        sd 0143 45
        ok 1
        ok 2
        ok 6

DESCRIPTION

    Serial console (19200) accepts commands and sets servo motor
    accordingly.

COMMANDS

    [X] is servo index 1-6 or bitwise 000-077 (6 last bits as mask),
    default is zero -> 077

 sd [X] [int] [int]

    Set servo angle. Value range 0 to 180.

 ss [X] [int] [int]

    Set servo angle using PWM peak time length in nano seconds. Value range
    500 to 2500.

sweep [X] [int] [int]

    Moves servo with index of first parameter from 90 to 180, then back to
    0 and again to 90 degree.

    Both parameters are optional. First is the number of repeats, default
    1. Second sets the time in milliseconds to wait between each angle
    change, default 15.

d [X]

    Will detach - stop sending PWM to servo, turning the control off.

z [X]

    Will set current servo position as zero.

c [X]

    Will calibrate servo -> calculate feedback be moving servo form current
    position back and forth.

m [X] [int] [int]

    Will monitor servos and return sr commands based on the feedback pin
    readings.

    Both paramters are optional. First parameter is the count of how many
    readings will be done. Use -1 or 0xffff to set infinite count. Will
    finish on any serial input. Second parameter is the delay between
    reading.

dump [X]

    Will dump servo data.

save [X]

    Will save servo data into eeprom for permanent storage.

rfst [X]

    Factory-reset will clear eeprom and start from defaults.

SEE ALSO

    KiCAD PCBs are here:
    https://github.com/jozef/sixservo/tree/master/KiCAD, first single servo
    version: https://blog.kutej.net/2020/04/servo-tester

LICENSE

    This is free software, licensed under the MIT License.



Changelog
0.01 2020-10-25
    - initial version



singleservo.h
#ifndef singleservo_h
#define singleservo_h

#include <Arduino.h>
#include <Servo.h>

#define SPOS_MIN 500
#define SPOS_MAX 2500

struct singleservo_data {
    uint8_t signal_pin;
    uint8_t feedback_pin;

    bool is_calibrated = false;
    bool is_clockwise  = true;
    bool is_disabled   = false;

    uint16_t spos_zero = 1500;
    uint16_t spos_min  = SPOS_MIN;
    uint16_t spos_max  = SPOS_MAX;
    int16_t pos_min   = -500;
    int16_t pos_max   =  500;
    uint16_t fpos_zero =  511;
    uint16_t fpos_min  =    0;
    uint16_t fpos_max  = 1023;
};

class singleservo {
    public:
        singleservo(uint8_t spin, uint8_t fpin);
        singleservo(singleservo_data init_d);

        singleservo_data d;
        bool has_feedback;

        bool set_rpos(int16_t new_pos);
        bool set_rfpos(int16_t new_pos, uint16_t in_millis);
        bool set_rpos(int16_t new_pos, bool wait_stable);
        int16_t get_rpos();
        bool set_zero_pos_current();

        bool set_spos(uint16_t new_spos);
        bool set_spos(uint16_t new_spos, bool do_wait_stable);
        uint16_t get_spos();
        uint16_t get_fpos();
        uint16_t get_fpos(uint8_t scount);
        bool set_dpos(uint8_t new_dpos);
        bool set_dpos(uint8_t new_dpos, bool do_wait_stable);

        bool calibrate();
        void detach();
        bool wait_stable();
        void dump_data();
        void tick();
    private:
        uint16_t _last_spos = 0;
        uint16_t _last_fpos = 0;

        // future/tick positions
        int16_t _t_slots = 0;
        uint16_t _next_spos = 0;
        unsigned long _next_spos_millis = 0;
        uint16_t _next_spos2 = 0;
        unsigned long _next_spos2_millis = 0;
        Servo _servo;

        void _init();
};

#endif


singleservo.cpp
#include <Arduino.h>

#include "singleservo.h"
#include <Servo.h>

#define WAIT_CHANGE_MS 100   // how long to wait for feedback to stabilize
#define MAX_ITER 300
#define CALIB_RANGE 0x160
#define NC_TH 5
#define DEBUG 0
#define TICK_MS 30

void singleservo::detach() {
    if (_servo.attached()) {
        _servo.detach();
    }
}

bool singleservo::set_dpos(uint8_t new_dpos) {
    return set_dpos(new_dpos, false);
}

bool singleservo::set_dpos(uint8_t new_dpos, bool do_wait_stable) {
    if (new_dpos > 180) return false;

    return set_spos(
        map(new_dpos, 0, 180, d.spos_min, d.spos_max),
        do_wait_stable
    );
}

bool singleservo::set_spos(uint16_t new_spos) {
    return set_spos(new_spos, false);
}

bool singleservo::set_spos(uint16_t new_spos, bool do_wait_stable) {
    if (d.is_disabled) return false;
    if (new_spos < d.spos_min) {
        set_spos(d.spos_min, do_wait_stable);
        return false;
    }
    if (new_spos > d.spos_max) {
        set_spos(d.spos_max, do_wait_stable);
        return false;
    }

    if (!_servo.attached()) {
        _servo.attach(d.signal_pin, d.spos_min, d.spos_max);
    }
    _servo.writeMicroseconds(new_spos);
    _last_spos = new_spos;

    return (do_wait_stable ? wait_stable() : true);
}

bool singleservo::set_rpos(int16_t new_rpos) {
    return set_rpos(new_rpos, false);
}

bool singleservo::set_rpos(int16_t new_rpos, bool do_wait_stable) {
    if (d.is_clockwise) new_rpos *= -1;
    return set_spos(d.spos_zero + new_rpos, do_wait_stable);
}

bool singleservo::wait_stable() {
    if (!has_feedback) {
        return false;
    }

    uint16_t max_iterations = MAX_ITER + 1;
    while (--max_iterations) {
        uint16_t _prev_fpos = get_fpos();
        delay(WAIT_CHANGE_MS);

        get_fpos();
        if (_prev_fpos == _last_fpos) {
            if (DEBUG) {
                Serial.print(F("stable at fpos: "));
                Serial.println(_last_fpos);
            }
            break;
        }
    }

    return (max_iterations ? true : false);
}

uint16_t singleservo::get_fpos() {
    return get_fpos(4);
}

uint16_t singleservo::get_fpos(uint8_t scount) {
    uint16_t samples = 0;
    for (int8_t i = 0; i < scount; i++) {
        if (i) delay(10);
        samples += analogRead(d.feedback_pin);
    }
    _last_fpos = samples / scount;
    return _last_fpos;
}


void singleservo::dump_data() {
    Serial.print(F("signal pin: "));    Serial.println(d.signal_pin);
    Serial.print(F("feedback pin: "));  Serial.println(d.feedback_pin);
    Serial.print(F("has feedback: "));  Serial.println(has_feedback);
    Serial.print(F("is calibrated: ")); Serial.println(d.is_calibrated);
    Serial.print(F("is clockwise: "));   Serial.println(d.is_clockwise);
    Serial.print(F("is disabled: "));   Serial.println(d.is_disabled);
    Serial.print(F("spos zero: "));     Serial.println(d.spos_zero);
    Serial.print(F("spos min: "));      Serial.println(d.spos_min);
    Serial.print(F("spos max: "));      Serial.println(d.spos_max);
    Serial.print(F("pos min: "));       Serial.println(d.pos_min);
    Serial.print(F("pos max: "));       Serial.println(d.pos_max);
    Serial.print(F("fpos zero: "));     Serial.println(d.fpos_zero);
    Serial.print(F("fpos min: "));      Serial.println(d.fpos_min);
    Serial.print(F("fpos max: "));      Serial.println(d.fpos_max);
    Serial.print(F("fpos cur: "));      Serial.println(get_fpos());
}

void singleservo::_init() {
    pinMode(d.feedback_pin, INPUT);
    has_feedback = (get_fpos(8) < NC_TH ? false : true);
    detach();
}

singleservo::singleservo(uint8_t spin, uint8_t fpin) {
    d.signal_pin = spin;
    d.feedback_pin = fpin;
    _init();
}

singleservo::singleservo(singleservo_data init_d) {
    d = init_d;
    _init();
}

bool singleservo::set_zero_pos_current() {
    if (!has_feedback) {
        if (_last_spos == 0) return false;

        set_spos(_last_spos);
        d.spos_zero = _last_spos;
        return true;
    }

    uint16_t cur_spos = (_last_spos > 0 ? _last_spos : d.spos_zero);
    uint16_t wanted_fpos = get_fpos();

    uint16_t max_iterations = MAX_ITER + 1;
    while (--max_iterations) {
        set_spos(cur_spos);
        delay(WAIT_CHANGE_MS);

        get_fpos();
        if (wanted_fpos == _last_fpos) {
            break;
        }
        else {
            if (wanted_fpos > _last_fpos) {
                cur_spos++;
            }
            else {
                cur_spos--;
            }
            if (abs(wanted_fpos - _last_fpos) > 10) {
                cur_spos += ((wanted_fpos - _last_fpos) * 2);
            }
        }
    }

    if (max_iterations != 0) {
        d.spos_zero = cur_spos;
        d.fpos_zero = wanted_fpos;
        return true;
    }
    else {
        return false;
    }
}

int16_t _approx_border (int16_t l10, int16_t l11, int16_t l12, int16_t l20, int16_t l21) {
    int32_t ll11 = l11 - l10;
    int32_t ll12 = l12 - l11;
    int32_t ll21 = l21 - l20;
    int32_t ll22 = (ll21 * ll12) / ll11;
    return (l21 + ll22);
}

bool singleservo::calibrate() {
    if (!set_zero_pos_current()) {
        return false;
    }

    set_rpos(CALIB_RANGE, true);
    delay(3*WAIT_CHANGE_MS);
    uint16_t fpos_cw = get_fpos(20);
    uint16_t spos_cw = _last_spos;
    d.fpos_min = _approx_border(d.spos_zero, spos_cw, d.spos_min, d.fpos_zero, fpos_cw);

    set_rpos(0-CALIB_RANGE, true);
    delay(3*WAIT_CHANGE_MS);
    uint16_t fpos_ccw = get_fpos(20);
    uint16_t spos_ccw = _last_spos;
    d.fpos_max = _approx_border(d.spos_zero, spos_ccw, d.spos_max, d.fpos_zero, fpos_ccw);

    set_rpos(0);
    d.is_calibrated = true;

    return true;
}

uint16_t singleservo::get_spos() {
    if (!has_feedback) return _last_spos;

    return map(get_fpos(), d.fpos_min, d.fpos_max, d.spos_min, d.spos_max);
}

int16_t singleservo::get_rpos() {
    if (!d.is_calibrated) return 0;

    return (d.is_clockwise ? 1 : -1) * (d.spos_zero - get_spos());
}

bool singleservo::set_rfpos(int16_t new_rpos, uint16_t in_millis) {
    if (_last_spos == 0 ) {
        _last_spos = get_spos();
        if (_last_spos == 0 ) {
            set_rpos(new_rpos);
        }
    }

    if (d.is_clockwise) new_rpos *= -1;
    if (_next_spos) {
        _next_spos2 = d.spos_zero + new_rpos;
        _next_spos2_millis = _next_spos_millis + in_millis;
    }
    else {
        _next_spos = d.spos_zero + new_rpos;
        _next_spos_millis = millis() + in_millis;
        _t_slots = in_millis / TICK_MS;
    }

    return true;
}

void singleservo::tick() {
    if (!_next_spos_millis) return;

    long diff_t = _next_spos_millis - millis();
    int16_t slots = diff_t / TICK_MS;
    if ((diff_t <= 0) || (slots < 1)) {
        set_spos(_next_spos);
        _next_spos = _next_spos2;
        _next_spos_millis = _next_spos2_millis;
        _t_slots = (_next_spos_millis - millis()) / TICK_MS;
        _next_spos2 = 0;
        _next_spos2_millis = 0;
    }
    else {
        if (slots == _t_slots) {
            return;
        }
        int16_t diff_s = (_next_spos - _last_spos);
        set_spos( _last_spos + (diff_s / slots) );
        _t_slots = slots;
    }
}


examples/sixservo-mirror.pl
#!/usr/bin/env perl

use strict;
use warnings;
use 5.010;

use AnyEvent;
use AnyEvent::SerialPort;
use AnyEvent::IO qw(aio_open aio_write :flags);
use Getopt::Long;
use Pod::Usage;

my $MONITOR_PORT = '/dev/ttyUSB1';
my $ACTOR_PORT   = '/dev/ttyUSB0';

exit main();

sub main {
    my ($help,$do_log);
    GetOptions(
        'log_on_enter=s' => \$do_log,
        'help|h'         => \$help,
    ) or pod2usage(1);
    pod2usage(0) if $help;


    my $run_cv = AnyEvent->condvar;

    my $actor_ae = AnyEvent::SerialPort->new(
        serial_port => [
            $ACTOR_PORT,
            [baudrate => 19200],
            [parity   => 'none'],
            [databits => 8],
            [stopbits => 1],
        ],
        on_error => sub {
            my ($hdl, $fatal, $msg) = @_;
            die $msg;
        },
        on_eof => sub {
            $run_cv->send;
        },
    );
    $actor_ae->on_read(
        sub {
            my ($hdl) = @_;
            $hdl->push_read(
                line => sub {
                    my ($hdl, $line) = @_;
                    say '< a: ' . $line;
                }
            );
        }
    );

    my $monitor_ae = AnyEvent::SerialPort->new(
        serial_port => [
            $MONITOR_PORT,
            [baudrate => 19200],
            [parity   => 'none'],
            [databits => 8],
            [stopbits => 1],
        ],
        on_error => sub {
            my ($hdl, $fatal, $msg) = @_;
            die $msg;
        },
        on_eof => sub {
            $run_cv->send;
        },
    );
    my %mon_servos;
    $monitor_ae->on_read(
        sub {
            my ($hdl) = @_;
            $hdl->push_read(
                line => sub {
                    my ($hdl, $line) = @_;
                    say '< m: ' . $line;
                    if ($line =~ m/^srf? (\d) (-?\d+)\b/) {
                        $mon_servos{$1} = $2;
                        say '> a: ' . $line;
                        $actor_ae->push_write($line . "\n");
                    }
                }
            );
        }
    );
    my $posponed_start = AE::timer(
        3, 0,
        sub {
            $monitor_ae->push_write("m 077 -1 1\n");
        }
    );

    my $term_cb = sub {
        say 'terminating, bye bye';
        $monitor_ae->push_write("d 077\n");
        $actor_ae->push_write("d 077\n");
        my $w;
        $w = AE::timer(1, 0, sub {$w = undef; $run_cv->send;})
    };
    my $wt = AE::signal TERM => $term_cb;
    my $wh = AE::signal HUP  => $term_cb;
    my $wi = AE::signal INT  => $term_cb;

    my $log_fh;
    my $stdin_ae;
    if ($do_log) {
        aio_open(
            $do_log,
            (O_WRONLY | O_CREAT | O_APPEND),
            0600,
            sub {
                ($log_fh) = @_ or return AE::log error => "$! - can not open log";
            }
        );

        $stdin_ae = AnyEvent->io (fh => \*STDIN, poll => 'r', cb => sub {
            <STDIN>;
            return unless $log_fh;
            return unless %mon_servos;

            foreach my $servo_no (sort keys %mon_servos) {
                aio_write($log_fh, sprintf("sr %d %d\n", $servo_no, $mon_servos{$servo_no}), sub {});
            }
            aio_write($log_fh, "# ---\n", sub {});
            say "--- saved ---\n";
        });
    }

    $run_cv->recv;

    return 0;
}


__END__

=head1 NAME

sixservo-mirror.pl - will replay monitored servo moves

=head1 SYNOPSIS

    sixservo-mirror.pl
        --log_on_enter=FILE     will save current C<sr> values into a FILE
        --help                  print out this help

=head1 DESCRIPTION

Will turn on monitoring on F</dev/ttyUSB1> and send them to F</dev/ttyUSB0>.

=head1 LINKS

See script in action: L<https://www.instagram.com/p/CF66RVUDJ8H/>

=cut


examples/sixservo-replay.pl
#!/usr/bin/env perl

use strict;
use warnings;
use 5.010;

use AnyEvent;
use AnyEvent::SerialPort;
use Path::Class qw(file);
use Time::HiRes qw(time);
use Getopt::Long;
use Pod::Usage;

my $DEFAULT_S_PORT = '/dev/ttyUSB0';
my $STEP_MS_FIRST  = 3000;
my $STEP_MS_BASE   = 750;

exit main();

sub main {
    my ($help,$do_init,@rigid, @serial_ports);
    my $step_ms      = $STEP_MS_BASE;
    GetOptions(
        'help|h' => \$help,
        'port=s' => \@serial_ports,
        'step=s' => \$step_ms,
        'rigid'  => \@rigid,
        'init'   => \$do_init,
    ) or pod2usage(1);
    pod2usage(0) if $help;
    @serial_ports = ($DEFAULT_S_PORT)
        unless @serial_ports;

    # read/update commands
    my @cmds;
    my @last_cmd;
    while (my $line = <>) {
        chomp($line);
        if ($line =~ m/^#/) {
            if (@last_cmd) {
                push(@cmds, [@last_cmd]);
                @last_cmd = ();
            }
            last if $do_init;
            next;
        }
        push(@last_cmd, $line);
    }
    push(@cmds, \@last_cmd)
        if @last_cmd;
    exit(1) unless @cmds;

    # run
    my (@run_cvs, @all_timers);
    foreach my $serial_port (@serial_ports) {
        my ($run_cv, $timers) = schedule_cmds($serial_port, $step_ms, [@cmds], shift(@rigid));
        push(@run_cvs,    $run_cv);
        push(@all_timers, @$timers);
    }
    foreach my $run_cv (@run_cvs) {
        $run_cv->recv;
    }
    return 0;
}

sub schedule_cmds {
    my ($serial_port, $step_ms, $cmds, $rigid) = @_;

    my $run_cv = AnyEvent->condvar;

    my $actor_ae = AnyEvent::SerialPort->new(
        serial_port => [
            $serial_port,
            [baudrate => 19200],
            [parity   => 'none'],
            [databits => 8],
            [stopbits => 1],
        ],
        on_error => sub {
            my ($hdl, $fatal, $msg) = @_;
            die $msg;
        },
        on_eof => sub {
            $run_cv->send;
        },
    );
    $actor_ae->on_read(
        sub {
            my ($hdl) = @_;
            $hdl->push_read(
                line => sub {
                    my ($hdl, $line) = @_;
                    say $serial_port.'< a: ' . $line;
                }
            );
        }
    );

    my $term_cb = sub {
        say 'terminating, bye bye';
        $actor_ae->push_write("d 077\n");
        my $w;
        $w = AE::timer(1, 0, sub {$w = undef; $run_cv->send;});
    };
    my $wt = AE::signal TERM => $term_cb;
    my $wh = AE::signal HUP  => $term_cb;
    my $wi = AE::signal INT  => $term_cb;

    my $do_one_cmd = sub {
        my ($is_first) = @_;
        return $run_cv->send
            unless @$cmds;

        say 'time: ' . time();

        my @cmd_set = @{shift(@$cmds)};
        foreach my $cmd (@cmd_set) {
            if (!$rigid) {
                my $step_ms = ($is_first ? ($STEP_MS_FIRST + int($step_ms / 3)) : $step_ms);
                $cmd =~ s/^sr\s(.+)/srf $1 $step_ms/;
            }

            say $serial_port.'> a: ' . $cmd;
            $actor_ae->push_write($cmd . "\n");
        }
    };

    my $first_step;
    my $ticker;
    my $posponed_start = AE::timer(
        ($STEP_MS_FIRST / 1000),
        0,
        sub {
            $first_step = AE::timer(0, ($STEP_MS_FIRST / 1000), sub { $do_one_cmd->(1) });
            $ticker = AE::timer(($STEP_MS_FIRST / 1000), $step_ms / 1000, sub { $do_one_cmd->(0) });
        }
    );

    return ($run_cv, [$posponed_start, $first_step, $ticker]);
}

__END__

=head1 NAME

sixservo-replay.pl - will replay monitored servo moves

=head1 SYNOPSIS

    sixservo-replay.pl moves-file.log
        --port=XXX      set serial port, (ex. /dev/ttyUSB1)
                        can be present multiple times
        --step=AAA      set time in milli seconds for one step
        --init          perform only first step/position
        --rigid         will replay commands as they are, no smoothening
                        can be present multiple times, once per port
        --help          this help

=head1 DESCRIPTION

Will send commands from STDIN to F</dev/ttyUSB0>.

=head1 LINKS

See F<sixservo-mirror.pl> with C<--log_on_enter>.

More info: L<https://blog.kutej.net/2020/10/bob-303>

=cut
